### PR TITLE
BUG/TST: linear_spectrum_fitting function in test file

### PR DESCRIPTION
**Pull request by licode**
Fix a bug in test file due to API change on Friday.

### CODE REVIEW

**Comment by tacaswell** [comment]
What commit broke this?

**Comment by licode** [comment]
https://github.com/Nikea/scikit-xray/commit/bddfde865687466467c62379d3aa8e1373b3ced3

**Comment by licode** [comment]
Change area to None.

File: skxray/fitting/xrf_model.py [hunk 1]
```diff
@@ -1117,12 +1116,12 @@ def linear_spectrum_fitting(spectrum, params, elemental_lines=None,
         if np.sum(total_y[:, i]) == 0:
             continue
         result_dict.update({total_list[i]: total_y[:, i]})
-        area = 0
+        area = None
         if area_option:
             area = out[i]*element_area[total_list[i]]
```
Comment (tacaswell) [inline skxray/fitting/xrf_model.py]: This computation should be _very_ cheap.

File: skxray/fitting/xrf_model.py [hunk 2]
```diff
@@ -1117,12 +1116,12 @@ def linear_spectrum_fitting(spectrum, params, elemental_lines=None,
         if np.sum(total_y[:, i]) == 0:
```
Comment (tacaswell) [inline skxray/fitting/xrf_model.py]: You should use `for i, func in enumerate(total_list):` which will reduce the amount of slicing you have to do below.  With clever application of zip could probably get rid of the slicing all together.

File: skxray/fitting/xrf_model.py [hunk 3]
```diff
@@ -1117,12 +1116,12 @@ def linear_spectrum_fitting(spectrum, params, elemental_lines=None,
...
+        area = None
         if area_option:
             area = out[i]*element_area[total_list[i]]
         area_dict.update({total_list[i]: area})
     result_dict.update(background=bg)
-    bg = 0
+    bg = None
     if area_option:
         bg = np.sum(bg)
     area_dict.update(background=bg)
```
Comment (tacaswell) [inline skxray/fitting/xrf_model.py]: why are you using `update` instead of `area_dict['background'] = bg`?

File: skxray/fitting/xrf_model.py [hunk 4]
```diff
@@ -1117,12 +1116,12 @@ def linear_spectrum_fitting(spectrum, params, elemental_lines=None,
...
-        area = 0
+        area = None
         if area_option:
             area = out[i]*element_area[total_list[i]]
         area_dict.update({total_list[i]: area})
     result_dict.update(background=bg)
-    bg = 0
+    bg = None
     if area_option:
         bg = np.sum(bg)
```
Comment (tacaswell) [inline skxray/fitting/xrf_model.py]: This is a no-op as you re-assign bg to be `None` just above this.

File: skxray/fitting/xrf_model.py (hunk 1 above)
Comment (licode) [inline skxray/fitting/xrf_model.py]: This function is also used in single pixel fitting, so I prefer to reduce unnecessary calculation.

File: skxray/fitting/xrf_model.py (hunk 3 above)
Comment (licode) [inline skxray/fitting/xrf_model.py]: Any major difference?

Comment (tacaswell) [inline skxray/fitting/xrf_model.py]: I suspect that `[]` is marginally faster

File: skxray/fitting/xrf_model.py (hunk 1 above)
Comment (tacaswell) [inline skxray/fitting/xrf_model.py]: The `if` block also isn't free.

File: skxray/fitting/xrf_model.py (hunk 3 above)
Comment (licode) [inline skxray/fitting/xrf_model.py]: Ah, good to know. I sometimes feel confused in choosing which one to use.

File: skxray/fitting/xrf_model.py (hunk 1 above)
Comment (licode) [inline skxray/fitting/xrf_model.py]: OK, I will run some tests, then make the decision.

Comment (tacaswell) [inline skxray/fitting/xrf_model.py]: If I were going to optimize this function, this is not where I would start.

Also, don't start optimizing until you know a) that it matters b) where the bottle neck is.

**Comment by tacaswell** [comment]
If you are always going to reference `total_y` by extracting columns, I strongly suggest transposing it before working on it so that you can pull out rows instead.  The memory access patterns are much better.

**Comment by licode** [comment]
Good to know. Thanks. I will send the correction soon. 

BTW, why accessing row faster? Is it because the underlining pointer structure, such as pointer referencing to row first, then to each item of the row?

**Comment by tacaswell** [comment]
Yes, by default numpy arrays are stored in memory as row-major (but you can ask them to be column major if you want as well).

It does not matter too much for small arrays (when the whole thing fits in cache) but once you get bigger you can start to get awful page hits (as you need to go back to main memory for _every_ element rather than every row).  I also suspect that there is a way to vertorize this loop.

**Comment by licode** [comment]
Done. 

Persuaded by Data structure is more central to programming. :+1:  https://twitter.com/igrigorik/status/587446362880344064?s=15

File: skxray/fitting/xrf_model.py [hunk 5]
```diff
@@ -1102,30 +1100,23 @@ def linear_spectrum_fitting(spectrum, params, elemental_lines=None,
...
-        if np.sum(total_y[:, i]) == 0:
+    for i, func in enumerate(total_list):
+        if np.sum(total_y[i, :]) == 0:
             continue
-        result_dict.update({total_list[i]: total_y[:, i]})
-        area = 0
-        if area_option:
-            area = out[i]*element_area[total_list[i]]
+        result_dict.update({total_list[i]: total_y[i, :]})
+        area = out[i]*element_area[total_list[i]]
```
Comment (tacaswell) [inline skxray/fitting/xrf_model.py]: you don't need to slice into total_list, you have it as `func`

Comment (tacaswell) [inline skxray/fitting/xrf_model.py]:
```python
for name, y, _out in izip(total_list, total_y, out):
    if not np.sum(y):
        continue
    result_dict[name] = y
    aread_dict[name] = _out * element_area[name]
```

**Comment by licode** [comment]
Corrections added.

**Pull request merged**